compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -113,7 +121,218 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
+    auto maybeGenericOp = getGenericOpOrGeneralizeContraction(
+        rewriter, result.getOwner(), allowGeneralizing);
+    if (failed(maybeGenericOp)) {
+      return failure();
+    }
+    auto genericOp = maybeGenericOp.value();
+    result = genericOp->getOpResult(resultIndex);
+
+    ArrayRef<int64_t> perm = transposeOp.getPermutation();
+    auto invPerm = invertPermutationVector(perm);
+
+    OpOperand *init = genericOp.getDpsInitOperand(resultIndex);
+    SmallVector<Value> inits = genericOp.getDpsInits();
+    Value newInit = createTranspose(rewriter, init->get(), perm);
+    inits[resultIndex] = newInit;
+
+    SmallVector<Type> resultTypes(genericOp->getResultTypes());
+    resultTypes[resultIndex] = newInit.getType();
+
+    SmallVector<AffineMap> newIndexingMaps = genericOp.getIndexingMapsArray();
+    AffineMap resultMap =
+        newIndexingMaps[genericOp.getNumDpsInputs() + resultIndex];
+    SmallVector<AffineExpr> newExprs =
+        applyPermutation(resultMap.getResults(), perm);
+    AffineMap transposedMap =
+        AffineMap::get(resultMap.getNumDims(), resultMap.getNumSymbols(),
+                       newExprs, rewriter.getContext());
+    newIndexingMaps[genericOp.getNumDpsInputs() + resultIndex] = transposedMap;
+
+    auto newGenericOp = rewriter.create<linalg::GenericOp>(
+        genericOp.getLoc(), resultTypes, genericOp.getDpsInputs(), newInit,
+        newIndexingMaps, genericOp.getIteratorTypesArray(),
+        /*bodyBuild=*/nullptr, linalg::getPrunedAttributeList(genericOp));
+    rewriter.cloneRegionBefore(genericOp.getRegion(), newGenericOp.getRegion(),
+                               newGenericOp.getRegion().begin());
+
+    // Remap iteration space of the generic to be in order of the output.
+    SmallVector<unsigned int> interchange;
+    int64_t permIdx = 0;
+    for (int i = 0, e = transposedMap.getNumDims(); i < e; ++i) {
+      if (transposedMap.isFunctionOfDim(i)) {
+        interchange.push_back(
+            llvm::cast<AffineDimExpr>(transposedMap.getResult(permIdx))
+                .getPosition());
+        permIdx++;
+        continue;
+      }
+      interchange.push_back(i);
+    }
+    auto interchangedGenericOp =
+        linalg::interchangeGenericOp(rewriter, newGenericOp, interchange);
+    if (failed(interchangedGenericOp)) {
+      return failure();
+    }
+
+    // Replace the result of the transpose with the transposed init.
+    rewriter.replaceOp(transposeOp,
+                       interchangedGenericOp->getResult(resultIndex));
+    for (auto [oldRes, newRes] : llvm::zip_equal(
+             genericOp.getResults(), interchangedGenericOp->getResults())) {
+      if (oldRes.getResultNumber() == resultIndex)
+        continue;
+      rewriter.replaceAllUsesWith(oldRes, newRes);
+    }
+    return success();
+  }
+
+private:
+  bool allowGeneralizing = false;
+};
+
+// Bubbles a transpose through a tensor.collapse_shape.
+class BubbleTransposeThroughCollapseShape
```

Comment:
Hmmm, this logic is probably duplicated from MLIR for propagating collapse shapes past elementwise operations. Instead of reimplementing for specific case. maybe we can just expose the upstream method as a utility on `LinalgOp` and use that here?

Reply:
The reason I reimplemented it this way is because I want to leave it as a transpose. The upstream method will create a `linalg.generic` op. I could try to add a callback to the upstream utility to allow for providing a callback to construct the reshaped op?

Reply:
There was a change upstream that modified the corresponding collapse method to create a named op if the original operation was a named op...

Reply:
Has the patch for it landed yet? Looking at the upstream pattern I still see this: https://github.com/llvm/llvm-project/blob/26101e8c50e07ea53869b0c7f343d76bfed61b02/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L862

Reply:
Not for the expansion stuff. Its only updated for the collapse variant  https://github.com/llvm/llvm-project/blob/26101e8c50e07ea53869b0c7f343d76bfed61b02/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L1590

Reply:
I see, I think that logic only applies for cases like `linalg.add`/`linalg.sub` that don't need to update any indexing maps or fields. For `linalg.transpose` we need to update the `permutation` attribute. I can upstream this pattern as is though.